jwst/resample/resample.py

```diff
@@ -216,6 +215,57 @@ def blend_output_metadata(self, output_model):
             ignore=ignore_list
         )
 
+    def _get_intensity_scale(self, img):
+        """
+        Compute an intensity scale from the input and output pixel area.
+
+        For imaging data, the scaling is used to account for differences
+        between the nominal pixel area and the average pixel area for
+        the input data.
+
+        For spectral data, the scaling is used to account for flux
+        conservation with non-unity pixel scale ratios, when the
+        data units are flux density.
+
+        Parameters
+        ----------
+        img : DataModel
+            The input data model.
+
+        Returns
+        -------
+        iscale : float
+            The scale to apply to the input data before drizzling.
+        """
+        input_pixflux_area = img.meta.photometry.pixelarea_steradians
+        if input_pixflux_area:
+            if 'SPECTRAL' in img.meta.wcs.output_frame.axes_type:
+                # Use the nominal area as is
+                input_pixel_area = input_pixflux_area
+
+                # If input image is in flux density units, correct the
+                # flux for the user-specified change to the spatial dimension
+                if 'sr' not in str(img.meta.bunit_data).lower():
+                    input_pixel_area *= self.pscale_ratio
```

Comment:
This is what worries me most and that is why I did not touch spectral data in my original PR that introduced `iscale` correction. There are a couple of things going on here and I am not sure this is correct.

See https://github.com/spacetelescope/jwst/pull/7894#issuecomment-1968443459 (and also previous two comments) for what is the root-cause of the error in the _imaging_ data (pre- #7894). In any case, I believe in that case it is correct to take `iscale ~ sqrt(pixel area ratio)`.

For the spectral data, output image is a 2D image with wavelength along one axis and spatial coordinate along the other axis. So, `pscale_ratio` would (or should?) affect only this spatial axis but, unfortunately, due to how `drizzle` implementation assumes pixels are "square", (sampling along) both axes is affected. However, this kind of resampling should have been taken care of by the `drizzle`... It is strange that this correction is needed at all and that is what worries me. In addition, it is not clear why `input_pixel_area *= self.pscale_ratio` and not, i.e., `input_pixel_area *= self.pscale_ratio**2`.

I need to think about this more, especially to understand why is this need and particularly in the form `iscale ~ sqrt(pixel ratio)` (and not as `~ sqrt(pixel area ratio)`).

Reply:
Yes, there are a few things going on here!  I added some discussion in the ticket about the difference between the imaging and spectral case in the JP ticket, in case you haven't seen that yet.

In short, the imaging and spectral scaling are addressing two different issues.  Since imaging data is always in surface brightness units, it doesn't need a scale factor for flux conservation - the value per pixel is the same, regardless of how big the pixel is.  Spectral data is sometimes in flux density units, in which case, the value per spatial area needs to be conserved - this is why it needs scaling if the pixel area changes, depending on the data units.

Also, for spectral data, the nominal spatial area is pixel extent along the slit times the slit width.  The extent along the slit is accounted for in the WCS, but the slit width is not.  This is why the `pscale_ratio` is not squared: there is only one spatial extent that might change in the WCS, even though there are two spatial coordinates.